Event Grid Deserializing Events Redesign

kerri-lee opened the pull request:
- Removed `EventGridConsumer` class, replaced with `Parse`, `GetData<T>`, and `GetData` methods in `CloudEvent` and `EventGridEvent` types. This removes the need for users to register custom event types - now, they can simply deserialize to the type that want using `GetData<T>` or deserialize into system event types or `BinaryData` if the type is unknown using `GetData`.
- Added extension methods to `BinaryData` that will parse into a single event.
- Edited consuming event tests
- Edited live tests (due to making cloud event `Data` internal; also added additional testing for the new `BinaryData` constructor)
- Regenerated some models using updated swaggers

Commit:
initial changes to cloudevent

Commit:
implemented new prototype of cloudevent

Commit:
Added Parse methods to EGEvent and CloudEvent

Commit:
Updated implementation for lazy deserialization approach

(Warning: didn't delete EGConsumer/Options because the amount of errors in ConsumeEventTests would have been horrifying) Need to do: add ObjectSerializer param to Parse, Add back ObjectSerializer to PublisherClientOptions, Add BinaryData constructor overload to EGEvent/CloudEvent

Commit:
More changes to design

Removed serializer from event constructors, edited tests

Commit:
Added constructor overloads for binary data

Commit:
Updating API based on decisions made for consuming events

GetData() now returns system data type or BinaryData. Removed SystemData property. Deserializer moved to GetData<T>(). Edited tests for consuming events.

Commit:
Removed async methods for GetData when custom serializer is not passed in

Commit:
Small changes to EGEvent/CloudEvent, Edited tests

Commit:
Addressed cases involving primitive/string data

Commit:
Miscellaneous changes

Updated recorded tests after adding test for sending cloud events with BinaryData

Commit:
Added BinaryData overload to Parse method

Other code cleanup, transfer AdditionalProperties from internal cloud event to ExtensionAttributes dictionary in returned cloud event, GetData<T>() throws InvalidOperationException if event wasn't created from Parse() method

Commit:
Fixing documentation

Commit:
Added extension methods for BinaryData to parse events

Added tests for BinaryData.ToEventGridEvent/ToCloudEvent. Also, removed the ctor overload for BinaryData event data in EventGridEvent because we have to assume that the data is already serialized to JSON. This isn't a problem in CloudEvent because the BinaryData is sent as bytes.

Commit:
Edited consuming event tests

Commit:
Added more consumer tests

Commit:
Added data content type parameter to cloud event constructors

Media type (ex. pplication/json or pplication/xml) should be specified when sending data. If content type is empty, JSON value is implied.

Commit:
Added null check for data content type

Commit:
Addressing content type on live tests

(Immediately remembers to change something right after committing)

Commit:
Forgot to export API

Commit:
Attempting to fix compiler error

Error in ConsumeEventTests was: 'BinaryData' is a type, which is not valid in the given context

Commit:
Fixed binary data test for cloud event

Please don't fail me tests

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs:
```diff
@@ -24,15 +32,96 @@ public CloudEvent(string source, string type)
+        /// <param name="source"> Identifies the context in which an event happened. The combination of id and source must be unique for each distinct event. </param>
+        /// <param name="type"> Type of event related to the originating occurrence. For example, "Contoso.Items.ItemReceived". </param>
+        /// <param name="dataContentType"> Content type of the payload. A content type different from "application/json" should be specified when sending binary data. </param>
+        public CloudEvent(BinaryData data, string source, string type, string dataContentType)
```
`dataContentType` is optional so we should use an optional parameter.

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridEvent.cs:
```diff
@@ -47,5 +57,163 @@ public EventGridEvent(string subject, object data, string eventType, string data
+            }
+
+            // Reserialize JsonElement to stream
+            return BinaryData.FromStream(SerializePayloadToStream(SerializedData, new JsonObjectSerializer()));
```
Use the static JsonObjectSerializer instance once we add it.

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:
```diff
@@ -212,22 +216,11 @@ private async Task<Response> PublishCloudEventsInternal(IEnumerable<CloudEvent>
                     // The 'Data' property is optional for CloudEvents
```
nit: Not only is Data optional, but passing binary data into the ctor would result in this being null.

Commit:
Making reviewer changes

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs:
```diff
@@ -38,44 +37,15 @@ public CloudEvent(string source, string type)
-        /// <param name="type"> Type of event related to the originating occurrence. For example, "Contoso.Items.ItemReceived". </param>
         /// <param name="dataContentType"> Content type of the payload. A content type different from "application/json" should be specified if payload is not JSON. </param>
-        public CloudEvent(object data, string source, string type, string dataContentType)
+        public CloudEvent(string source, string type, object data, string dataContentType = null)
```
nit: use default

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:
```diff
@@ -120,11 +120,18 @@ private async Task<Response> SendEventsInternal(IEnumerable<EventGridEvent> even
-                        _dataSerializer.Serialize(stream, egEvent.Data, egEvent.Data.GetType(), cancellationToken);
-                        stream.Position = 0;
-                        data = JsonDocument.Parse(stream);
+                        data = JsonDocument.Parse(binaryEventData);
```
Is this using the implicit cast to ROM?

kerri-lee replied:
Yeah

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/tests/ConsumeEventTests.cs:
```diff
@@ -104,8 +104,14 @@ public void ConsumeEventUsingBinaryDataExtensionMethod()
+
             EventGridEvent egEvent = new EventGridEvent(new ContosoItemReceivedEventData(), "", "Contoso.Items.ItemReceived", "1");
-            Assert.That(() => egEvent.GetData<ContosoItemReceivedEventData>(),
+            Assert.That(() => egEvent.GetData<ContosoItemReceivedEventData>(customSerializer),
```
Can we also add in a GetData calls without the serializer (both generic and non-generic)?

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:
```diff
@@ -119,10 +119,20 @@ private async Task<Response> PublishEventsInternal(IEnumerable<EventGridEvent> e
-                    stream.Position = 0;
-                    JsonDocument data = JsonDocument.Parse(stream);
+                    JsonDocument data;
+                    if (egEvent.Data is BinaryData binaryEventData)
```
Can we add a test that covers this? The generated request should have Json of BinaryData, but rather the thing that BinaryData was wrapping.

JoshLove-msft replied:
Turns out BinaryData can't be serialized anyway.

Commit:
More reviewer changes

Adding extra tests for publishing event grid events with binary data, testing GetData

Commit:
export API

pakrym commented on sdk/eventgrid/Azure.Messaging.EventGrid/api/Azure.Messaging.EventGrid.netstandard2.0.cs:
```diff
@@ -1,9 +1,15 @@
+    public static partial class BinaryDataExtensions
+    {
+        public static Azure.Messaging.EventGrid.CloudEvent ToCloudEvent(this Azure.Core.BinaryData binaryData) { throw null; }
+        public static Azure.Messaging.EventGrid.EventGridEvent ToEventGridEvent(this Azure.Core.BinaryData binaryData) { throw null; }
```
Are we sure we want to start shipping extension methods for an exchange type like this?

JoshLove-msft replied:
@KrzysztofCwalina suggested adding this

pakrym replied:
VS shows extension methods from all referenced packages ignoring extensions.

![image](https://user-images.githubusercontent.com/1697911/91108788-c8807500-e62d-11ea-8a5e-d00f1e961e27.png)

Do we really want to make `BinaryData` an extension method dumping ground?

JoshLove-msft replied:
I guess something else to consider is that even when thinking strictly from service bus context, not every message will be convertible to an eg event. TryParse might be a better pattern here.

JoshLove-msft replied:
I guess something else to consider is that even when thinking strictly from service bus context, not every message will be convertible to an eg event. TryParse might be a better pattern here.

Commit:
Added logic for serializing BinaryData to json for eg events

Added test for sending non-serialized BinaryData - should serialize the string representation of BinaryData

JoshLove-msft commented on sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/EventGridPublisherClient.cs:
```diff
@@ -122,16 +122,18 @@ private async Task<Response> SendEventsInternal(IEnumerable<EventGridEvent> even
-                        data = JsonDocument.Parse(binaryEventData);
+                        try
+                        {
+                            data = JsonDocument.Parse(binaryEventData);
```
This somewhat highlights where a DataFormat property on BinaryData might help. We don't know whether the BinaryData is already serialized or not, so we have to do this try/catch.
